src/code/providers/s3-share-provider.ts

```diff
@@ -101,19 +102,39 @@ class S3ShareProvider implements IShareProvider  {
     callback: callbackSigShare) {
 
     // document ID is stored in masterContent
-    const documentID = masterContent.get('sharedDocumentId')
+    let documentID = masterContent.get('sharedDocumentId')
 
     // newer V2 documents have 'accessKeys'; legacy V1 documents have 'sharedEditKey's
     // which are actually V1 'runKey's under an assumed name (to protect their identity?)
     const accessKeys = masterContent.get('accessKeys')
     const runKey = masterContent.get('shareEditKey')
-    const accessKey = accessKeys?.readWrite || runKey
+    let readWriteToken = accessKeys?.readWrite || runKey
     const contentJson = sharedContent.getContentAsJSON()
     const filename = metadata.filename
     // if we already have a documentID and some form of accessKey,
     // then we must be updating an existing shared document
-    if (documentID && accessKey) {
-      this.updateShare(contentJson, documentID, filename, accessKey, callback)
+    if (documentID && readWriteToken) {
+      // There are two kinds of documents we might encounter:
+      // 1. Documents shared using old CFM based on document-store.
+      // 2. Documents shared using new, token-service based CFM.
+      // There are a few ways to recognize legacy documents that require special handling:
+      // - readWriteToken key doesn't start with "read-write-token" prefix
+      // - documentID is a number (new docs have Firestore IDs which are combinations of digits and characters)
+      // - there is no "sharedDocumentUrl"
+      // Any of these methods can be used to detect legacy document.
+      const isLegacyDocument = readWriteToken.indexOf("read-write-token") !== 0
+      if (isLegacyDocument) {
+        // This logic is based on logic in the document store migration script:
+        // https://github.com/concord-consortium/document-store/blob/master/token-service-migration/index.js#L115-L126
+        documentID = accessKeys.readOnly
+        if (!documentID) {
+          // document-store migration does the same thing if readOnly key is missing.
+          documentID = sha256(readWriteToken)
+        }
+        // Again, follow document-store migration code.
+        readWriteToken = "read-write-token:doc-store-imported:" + readWriteToken
```

Comment:
Based on your Slack comment this should be something like:

`("read-write-token:doc-store-imported:" + readWriteToken).substr(0, 64)`

or

`("read-write-token:imported:" + readWriteToken).substr(0, 64)`

Reply:
I don't think this error is about that. readWriteToken can't be truncated, as that's the key that token-service will compare with one saved in Firestore.

I believe the token-service error is coming from:
https://github.com/concord-consortium/token-service/blob/master/functions/src/base-resource-object.ts#L432
So I think that truncation needs to happen there.